### PR TITLE
Add bundle root for generation

### PR DESCRIPTION
**Pull request by sfc-gh-bdufour**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description

Introducing `bundle_root` as a new project property for native apps. The bundle root is similar to the deploy root, but is going to be used during code generation to store temporary artifacts. For example, an upcoming feature will require setting up a virtual environment for the user so that Python code can be appropriately sandboxed. That environment will be created in the bundle root.

The bundle root change it itself was very small. Most of this PR actually involves wiring the bundle root through to the compiler. Instead of adding one more parameter to the compiler and processors, I felt it was time to take a step change and recheck the contract to the processors. So I extracted the data portion of the native app manager, and am I using it as a kind of "preserve whole object" refactoring. This way, processors can access the information they need without having to thread it through the set of calls every time we need a new piece of data.

I have avoided updating the release notes for now, as this should be part of the actual feature entry when it gets released. The bundle root is, at the moment, unused by any logic, so it doesn't make sense to announce it as something new by itself.

### Testing

* Executed the manual test plan to make sure there were no regressions
* Added unit tests tests to reflect the change in APIs. Integ and E2E tests are unaffected by the refactoring
* Manually checked for regressions in the code generation feature

### CODE REVIEW

File: src/snowflake/cli/plugins/nativeapp/project_model.py
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/plugins/nativeapp/project_model.py:59]: All of these were just moved verbatim from the manager class

File: src/snowflake/cli/plugins/nativeapp/project_model.py
Comment (sfc-gh-bgoel) [inline src/snowflake/cli/plugins/nativeapp/project_model.py:37]: I believe this can be used incorporated in https://github.com/snowflakedb/snowflake-cli/blob/851bfdbda0cb4a9521c46fc4c17273adb36fbd36/src/snowflake/cli/api/sql_execution.py#L91 to reduce the duplicate logic in that context manager.

File: src/snowflake/cli/plugins/nativeapp/manager.py
Comment (sfc-gh-bgoel) [inline src/snowflake/cli/plugins/nativeapp/manager.py:166]: nit: Since we're renaming some stuff, we can make this explicit that this is not the entire project, but only the `na_project`

File: src/snowflake/cli/plugins/nativeapp/codegen/artifact_processor.py
Comment (sfc-gh-bgoel) [inline src/snowflake/cli/plugins/nativeapp/codegen/artifact_processor.py:50]: I saw that the asserts have been moved from here to the SnowparkAnnotationProcessor class. Do we not want to have this assert at the base class level?

**Review by sfc-gh-bgoel — COMMENTED** [review]
1. I don't see the use of `bundle_root` yet, except to define its property method. I am guessing this is intentional and will be used with the new feature's code.
2. I am not sure I understand the example supplied for `bundle_root`, let's catch up offline to dive into it.

**Comment by sfc-gh-bdufour** [comment]
> 1. I don't see the use of `bundle_root` yet, except to define its property method. I am guessing this is intentional and will be used with the new feature's code.
> 2. I am not sure I understand the example supplied for `bundle_root`, let's catch up offline to dive into it.

1. Correct, that's all explained in the PR description. I'm already building on top of this PR.
2. What example are you referring to?

File: src/snowflake/cli/plugins/nativeapp/codegen/artifact_processor.py
Comment (sfc-gh-bdufour) [reply src/snowflake/cli/plugins/nativeapp/codegen/artifact_processor.py:50]: So far the Snowpark processor relies heavily on that invariant, but I didn't think it was a property of _all_ processors. In any case, I made it a strong invariant now, so the assertions are really redundant.

File: src/snowflake/cli/plugins/nativeapp/project_model.py
Comment (sfc-gh-bdufour) [reply src/snowflake/cli/plugins/nativeapp/project_model.py:37]: Spotted that too, but we don't have ownership where it should go, so I'd prefer to split it into its own submission to avoid noise for the CLI team.

**Review by sfc-gh-bgoel — DISMISSED** [review]
Caught up offline to go over the `bundle_root` introduction in more detail. The new addition + refactor LGTM. Acking failing IT.

File: src/snowflake/cli/api/project/schemas/native_app/native_app.py
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/native_app/native_app.py:37]: I'm not 100% sold on bundle root, but I can't think of a better name right now...

File: src/snowflake/cli/plugins/nativeapp/project_model.py
Comment (sfc-gh-cgorrie) [reply src/snowflake/cli/plugins/nativeapp/project_model.py:37]: TBH I think it's worth getting their sign-off here, it's so generally useful. But we can always make a one-off PR afterwards that moves it.

**Review by sfc-gh-cgorrie — APPROVED** [review]
Approved though I'm not completely sold on the name.

File: src/snowflake/cli/plugins/nativeapp/project_model.py
Comment (sfc-gh-bdufour) [reply src/snowflake/cli/plugins/nativeapp/project_model.py:37]: Yeah, I don't think the current_role() calls belongs there, I was always planning to move it due to it being generally useful. There's a location that has all these default wh/role/etc helpers, I'm planning to move it there too.

**Review by sfc-gh-fcampbell — APPROVED** [review]
LGTM pending integration test success

**Pull request merged**
